### PR TITLE
Requests: Add test for bundle

### PR DESCRIPTION
**Pull request by fitzthum**
Also allow keysets/bundles with no policy.

Signed-off-by: Tobin Feldman-Fitzthum <tobin@ibm.com>

### CODE REVIEW

File: db.sql [hunk 1]
```diff
@@ -52,7 +52,7 @@ CREATE TABLE `keysets` (
   `id` int(11) NOT NULL AUTO_INCREMENT,
   `keysetid` varchar(50) NOT NULL,
   `kskeys` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
-  `polid` int(11) NOT NULL,
+  `polid` int(11) DEFAULT NULL,
```
Comment (jodh-intel) [inline db.sql:55]: - Why is this change required?
- What happens to the rows without a `polid`? Won't they just grow forever, consuming more space that isn't strictly necessary?
- Please could you split this PR into two commits, adding some detail on the rationale for the `polid` change.

File: src/db.rs [hunk 2]
```diff
@@ -194,6 +194,22 @@ pub fn insert_keyset(ksetid: &str, kskeys: &[String], polid: u32) -> Result<()>
...
+
+    let mysqlstr = "INSERT INTO keysets (keysetid, kskeys)
+                    VALUES(?, ?)"
+        .to_string();
+
+    /* create JSON for vector struct member variables */
+    let kskeys_str = serde_json::to_string(kskeys)?;
+
+    trnsx.exec_drop(mysqlstr, (ksetid, &kskeys_str))?;
+    trnsx.commit()?;
```
Comment (jodh-intel) [inline src/db.rs:209]: This takes me back to my days working with the X/Open XA spec ;)

File: db.sql (hunk 1 above)
Comment (fitzthum) [reply db.sql:55]: Policies are optional for all entities. There is one default policy for the entire KBS and then any policies that are added for a keyset are added on top of that and any policies that are added for individual keys are added on top of that. This is why `get_secret_policy` and `get_keyset_policy` return Options rather than Results. No design change here we'd just never tested a keyset without a policy. 

I'm not sure about your second point. Keysets should be added or removed by the guest owner. I can put the db changes in a separate commit.

File: src/db.rs [hunk 3]
```diff
@@ -194,6 +194,22 @@ pub fn insert_keyset(ksetid: &str, kskeys: &[String], polid: u32) -> Result<()>
...
 
+pub fn insert_keyset_only(ksetid: &str, kskeys: &[String]) -> Result<()> {
+    let mut dbconn = get_dbconn()?;
+    let mut trnsx = dbconn.start_transaction(TxOpts::default())?;
+
+    let mysqlstr = "INSERT INTO keysets (keysetid, kskeys)
+                    VALUES(?, ?)"
+        .to_string();
+
+    /* create JSON for vector struct member variables */
```
Comment (dubek) [inline src/db.rs:205]: style: use `//` comments

File: src/request.rs [hunk 4]
```diff
@@ -159,11 +159,16 @@ impl SecretType for SecretBundle {
     }
 
     fn policies(&self) -> Vec<policy::Policy> {
-        let mut policies = vec![db::get_keyset_policy(&self.request.id).unwrap()];
+        let mut policies = vec![];
+        if let std::result::Result::Ok(keyset_policy) = db::get_keyset_policy(&self.request.id) {
+            policies.push(keyset_policy);
+        }
```
Comment (dubek) [inline src/request.rs:165]: 1. wouldn't it be semantically more accurate for `get_keyset_policy()` to return `Result<Option<Policy>>` -- Err if some DB stuff went wrong, Ok(None) if no policy, and Ok(Some(policy)) if the keyset-policy is found. I think that here a DB-error (say connection issues) will be treated as if there are no policies. (BTW the return type of `exec_first` is `Result<Option<T>>`)
2. Can we use the shorter `Ok(keyset_policy)` instead of `std::result::Result::Ok(keyset_policy)` ?

Comment (fitzthum) [reply src/request.rs:165]: Yeah this is probably a good idea. It makes sense to match up with the signature of the mysql methods. The only thing to consider is that most database errors probably shouldn't be exposed to the KBC. I'll look into it tomorrow.

Comment (fitzthum) [reply src/request.rs]: Ok, I added another commit to redo some of the error handling for `get_secret_policy` and `get_keyset_policy`. It's possible there is a more clever way to handle the `Result<Option<Policy>>` stuff, but I think this is decent. There are a few more tests and tweaks I want to do for `db.rs` but I'll probably leave that for another PR.

**Review by dubek — APPROVED** [review]
OK looks good.

One thing to consider (for the future) is to improve DB queries; for example, instead of selecting `polid` and then retrieving the policy details, you can do it in one query with a JOIN. Of course IMO performance is not the main issue now -- so code and execution flow being clear is more important.

File: src/request.rs [hunk 5]
```diff
@@ -205,6 +245,39 @@ mod tests {
...
+        let bundle_id = Uuid::new_v4().to_hyphenated().to_string();
+        let guid = "2cf13667-ea72-4013-9dd6-155e89c5a28f".to_string();
+        let request = RequestDetails {
+            guid: guid.clone(),
+            format: "json".to_string(),
+            secret_type: "bundle".to_string(),
+            id: bundle_id.clone(),
+        };
+
+        let secret_value = "dGVzdCBzZWNyZXQ=".to_string(); // "test secret" -> b64
```
Comment (jodh-intel) [inline src/request.rs:260]: Nit: I'd consider doing this "long hand" for clarity. It also means you can drop the comment which cannot get out of sync with the code:

```suggestion
        let secret = "test secret";
        let secret_value = base64::encode(secret);
```

**Review by jodh-intel — APPROVED** [review]
Thanks @fitzthum. A few nits but overall...

lgtm
